### PR TITLE
Stream aware outputs

### PR DESCRIPTION
**Pull request by mzient**
## Category:
**New feature** (*non-breaking change which adds functionality*)
**Refactoring** (*Redesign of existing code that doesn't affect functionality*)

## Description:
This PR adds support for returning the pipeline outputs as DLPack without copying.

### Affected modules and functionalities:
This PR adds the following features:
- stream parameters for Pipeline Run/Outputs
- `__dlpack__` interface for Tensors
It removes the `_expose_dlpack_capsule` as it was incomplete.

### Tests:

- [ ] Existing tests apply
- [X] New tests added
  - [X] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [X] Documentation updated
  - [X] Docstring
  - [X] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [X] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-4075

### CODE REVIEW

File: dali/test/python/dlpack/test_torch.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+    s = torch.cuda.Stream(0)
+    with torch.cuda.stream(s):
+        niter = 5
+        means = torch.zeros((niter * pipe.max_batch_size,), dtype=torch.float32).cuda()
+        flat_idx = 0
+        for i in range(niter):
+            (out,) = pipe.run(s)
+            # convert the tensors in the batch to DLPack
+            batch = [torch.from_dlpack(t) for t in out]
```
Comment (szkarpinski) [inline dali/test/python/dlpack/test_torch.py:33]: Nit: This comment is slightly confusing: "convert to DLPack", but the function is `from_dlpack`. Maybe rephrase it a bit to indicate that we're not really converting to DLPack, but DALI->DLPack->Torch (without a copy)

File: dali/test/python/dlpack/test_torch.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+            for t in batch:
+                means[flat_idx] = torch.mean(t)
+                flat_idx += 1
+        # those are meant to overwrite the results if synchronization fails
```
Comment (szkarpinski) [inline dali/test/python/dlpack/test_torch.py:37]: maybe check that we're actually sharing the memory:
```
batch_a = [torch.from_dlpack(t) for t in out]
batch_b = [torch.from_dlpack(t) for t in out]
# now change batch_b and make sure that batch_a is changed as well
```

File: dali/test/python/jax_plugin/test_integration.py [hunk 3]
```diff
@@ -41,7 +41,7 @@ def test_dali_tensor_gpu_to_jax_array(dtype, shape, value):
     assert jax.numpy.array_equal(jax_array, jax.numpy.full(shape, value, dtype))
 
     # Make sure JAX array is backed by the GPU
-    assert jax_array.device() == jax.devices()[0]
+    assert jax_array.device == jax.devices()[0]
```
Comment (szkarpinski) [inline dali/test/python/jax_plugin/test_integration.py:44]: A breaking change?

File: dali/pipeline/data/dltensor.cc [hunk 4]
```diff
@@ -12,13 +12,102 @@
...
 #include <string>
+#include <thread>
 #include "dali/pipeline/data/dltensor.h"
 #include "dali/core/error_handling.h"
+#include "dali/core/mm/detail/aux_alloc.h"
 #include "dali/core/static_switch.h"
 
 namespace dali {
 
+class DLTensorGraveyard {
```
Comment (szkarpinski) [inline dali/pipeline/data/dltensor.cc:28]: Maybe add some docs why we need this and how it works?

File: dali/test/python/jax_plugin/test_integration.py (hunk 3 above)
Comment (mzient) [reply dali/test/python/jax_plugin/test_integration.py:44]: Slipped in... That's a breaking change... in JAX 0.4.31 :\

File: dali/pipeline/data/dltensor.cc (hunk 4 above)
Comment (mzient) [reply dali/pipeline/data/dltensor.cc:28]: Will do.

File: dali/test/python/dlpack/test_torch.py (hunk 2 above)
Comment (mzient) [reply dali/test/python/dlpack/test_torch.py:55]: Done.

File: dali/test/python/dlpack/test_torch.py (hunk 1 above)
Comment (mzient) [reply dali/test/python/dlpack/test_torch.py:51]: Done.

File: dali/pipeline/data/dltensor.cc (hunk 4 above)
Comment (mzient) [reply dali/pipeline/data/dltensor.cc:40]: Done.

**Review by banasraf — APPROVED** [review]

**Review by szkarpinski — APPROVED** [review]
